src/serializers/type_serializers/dataclass.rs

```diff
@@ -131,16 +132,30 @@ impl TypeSerializer for DataclassSerializer {
         exclude: Option<&PyAny>,
         extra: &Extra,
     ) -> PyResult<PyObject> {
-        let extra = Extra {
+        let dc_extra = Extra {
             model: Some(value),
             ..*extra
         };
-        if self.allow_value(value, &extra)? {
-            let inner_value = self.get_inner_value(value)?;
-            self.serializer.to_python(inner_value, include, exclude, &extra)
+        if self.allow_value(value, extra)? {
+            let py = value.py();
+            if let CombinedSerializer::Fields(ref fields_serializer) = *self.serializer {
+                let output_dict = fields_serializer.main_to_python(
+                    py,
+                    DataclassResultIterator::new(&self.fields, value),
+                    include,
+                    exclude,
+                    dc_extra,
+                )?;
+
+                fields_serializer.add_computed_fields_python(Some(value), output_dict, include, exclude, extra)?;
+                Ok(output_dict.into_py(py))
+            } else {
+                let inner_value = self.get_inner_value(value)?;
```

Comment:
keep the old logic in case there are schemas with a dataclass serializer wrapping another serializer.